services/galley/src/Galley/API/Teams.hs

```diff
@@ -237,27 +238,59 @@ uncheckedAddTeamMember (tid ::: req ::: _) = do
 updateTeamMember (zusr::: zcon ::: tid ::: req ::: _) = do
-    body <- fromBody req invalidPayload
-    let user = body^.ntmNewTeamMember.userId
-    let perm = body^.ntmNewTeamMember.permissions
+    -- the team member to be updated
+    targetMember <- (view ntmNewTeamMember) <$> fromBody req invalidPayload
+    let targetId          = targetMember^.userId
+        targetPermissions = targetMember^.permissions
+
+    -- get the team and verify permissions
+    team    <- tdTeam <$> (Data.team tid >>= ifNothing teamNotFound)
     members <- Data.teamMembers tid
-    member  <- permissionCheck zusr SetMemberPermissions members
-    unless ((perm^.self) `Set.isSubsetOf` (member^.permissions.copy)) $
-        throwM invalidPermissions
-    unless (isTeamMember user members) $
-        throwM teamMemberNotFound
-    when (user `isOnlyOwner` members && perm /= fullPermissions) $
+    user    <- permissionCheck zusr SetMemberPermissions members
+
+    -- user may not elevate permissions
+    targetPermissions `ensureNotElevated` user
+
+    -- target user must be in same team
+    unless (isTeamMember targetId members) $
+      throwM teamMemberNotFound
+
+    -- cannot demote only owner (effectively removing the last owner)
+    when (targetId `isOnlyOwner` members
+          && targetPermissions /= fullPermissions) $
         throwM noOtherOwner
-    Data.updateTeamMember tid user perm
-    team <- tdTeam <$> (Data.team tid >>= ifNothing teamNotFound)
-    when (team^.teamBinding == Binding) $
-        Journal.teamUpdate tid (body^.ntmNewTeamMember : filter (\u -> u^.userId /= user) members)
+
+    -- update target in Cassandra
+    Data.updateTeamMember tid targetId targetPermissions
+
+    let otherMembers = filter (\u -> u^.userId /= targetId) members
+        updatedMembers = targetMember : otherMembers
+
+    -- note the change in the journal
+    when (team^.teamBinding == Binding) $ Journal.teamUpdate tid updatedMembers
+
+    -- inform members of the team about the change
+    -- some (privileged) users will be informed about which change was applied
+    let privilege = flip hasPermission GetMemberPermissions
+        (privileged, unprivileged) = partition privilege updatedMembers
+        mkUpdate           = EdMemberUpdate targetId
+        privilegedUpdate   = Just $ mkUpdate $ Just targetPermissions
+        unPrivilegedUpdate = Just $ mkUpdate Nothing
+        privilegedRecipients   = membersToRecipients Nothing privileged
+        unPrivilegedRecipients = membersToRecipients Nothing unprivileged
+
     now <- liftIO getCurrentTime
-    let e = newEvent MemberUpdate tid now & eventData .~ Just (EdMemberUpdate user)
-    let r = list1 (userRecipient zusr) (membersToRecipients (Just zusr) members)
-    push1 $ newPush1 zusr (TeamEvent e) r & pushConn .~ Just zcon
+    let ePriv  = newEvent MemberUpdate tid now & eventData .~ privilegedUpdate
+        eUPriv = newEvent MemberUpdate tid now & eventData .~ unPrivilegedUpdate
```

Comment:
This could actually also be written as
```
let ePriv  = newEvent MemberUpdate tid now & eventData ?~ privilegedUpdate
     eUPriv = newEvent MemberUpdate tid now & eventData ?~ unPrivilegedUpdate
```
and you could then remove the `Just` from `privilegedUpdate` and `unPrivilegedUpdate `